File: backend/src/main/java/com/backend/ord/domain/entities/Game.java
package com.backend.ord.domain.entities;

import com.backend.ord.domain.entities.abstracts.EntityBase;
import com.backend.ord.enums.Game.GameStatus;
import com.backend.ord.enums.Game.GameType;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

@Entity
@Table(name = "games")
public class Game extends EntityBase {
    @NotNull
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @NotNull
    @Column(name = "final_score", nullable = false)
    private Integer finalScore;

    @NotNull
    @Column(name = "acquired_points", nullable = false)
    private Integer acquiredPoints;

    @Enumerated(EnumType.STRING)
    @Column(name = "type", columnDefinition = "game_type(0, 0) not null", nullable = false)
    private GameType type;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", columnDefinition = "game_status(0, 0) not null", nullable = false)
    private GameStatus status;

    public @NotNull User getUser() {
        return this.user;
    }

    public @NotNull Integer getFinalScore() {
        return this.finalScore;
    }

    public @NotNull Integer getAcquiredPoints() {
        return this.acquiredPoints;
    }

    public GameType getType() {
        return this.type;
    }

    public GameStatus getStatus() {
        return this.status;
    }

    public void setUser(@NotNull User user) {
        this.user = user;
    }

    public void setFinalScore(@NotNull Integer finalScore) {
        this.finalScore = finalScore;
    }

    public void setAcquiredPoints(@NotNull Integer acquiredPoints) {
        this.acquiredPoints = acquiredPoints;
    }

    public void setType(GameType type) {
        this.type = type;
    }

    public void setStatus(GameStatus status) {
        this.status = status;
    }
}

File: backend/src/main/kotlin/com/backend/ord/domain/entities/Game.kt
package com.backend.ord.domain.entities

import com.backend.ord.domain.entities.interfaces.IdentifiableUserResource
import com.backend.ord.enums.persistance.game.GameDifficulty
import com.backend.ord.enums.persistance.game.GameStatus
import com.backend.ord.enums.persistance.game.GameType
import com.backend.ord.enums.persistance.language.LanguageName
import jakarta.persistence.*
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.OnDelete
import org.hibernate.annotations.OnDeleteAction
import org.hibernate.annotations.UpdateTimestamp
import java.time.Instant
import java.util.*

@Entity
@Table(name = "games")
class Game(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    override var id: UUID = UUID.randomUUID(),

    @Column(name = "final_score", nullable = false)
    var finalScore: Int = 0,

    @Column(name = "duration", nullable = false)
    var duration: String = "00:00:00",

    @Enumerated(EnumType.STRING)
    @Column(name = "type", columnDefinition = "game_type(0, 0) not null", nullable = false)
    var type: GameType,

    @Enumerated(EnumType.STRING)
    @Column(name = "difficulty", columnDefinition = "game_difficulty(0, 0) not null", nullable = false)
    var difficulty: GameDifficulty,

    @Enumerated(EnumType.STRING)
    @Column(name = "status", columnDefinition = "game_status(0, 0) not null", nullable = false)
    var status: GameStatus = GameStatus.IN_PROGRESS,

    @Column(name = "language", columnDefinition = "language_name(0, 0) not null")
    @Enumerated(EnumType.STRING)
    var language: LanguageName,

    @Column(name = "instruction", nullable = false, columnDefinition = "json")
    var instruction: String,

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JoinColumn(name = "user_id", nullable = false)
    override var user: User,

    @Column(name = "user_id", insertable = false, updatable = false)
    var userId: UUID = user.id,

    @Column(name = "created_at", nullable = false, updatable = false)
    @CreationTimestamp
    var createdAt: Instant = Instant.now(),

    @Column(name = "updated_at", nullable = false)
    @UpdateTimestamp
    var updatedAt: Instant = Instant.now()
) : IdentifiableUserResource {
    @PostLoad
    fun populateUserId() {
        userId = user.id
    }
}